registry/storage/blobstore.go

```diff
@@ -2,132 +2,164 @@ package storage
+// Get implements the BlobReadService.Get call.
+func (bs *blobStore) Get(ctx context.Context, dgst digest.Digest) ([]byte, error) {
 	bp, err := bs.path(dgst)
 	if err != nil {
 		return nil, err
 	}
 
-	return bs.driver.GetContent(bs.ctx, bp)
-}
+	p, err := bs.driver.GetContent(ctx, bp)
+	if err != nil {
+		switch err.(type) {
+		case driver.PathNotFoundError:
+			return nil, distribution.ErrBlobUnknown
+		}
 
-// link links the path to the provided digest by writing the digest into the
-// target file.
-func (bs *blobStore) link(path string, dgst digest.Digest) error {
-	if exists, err := bs.exists(dgst); err != nil {
-		return err
-	} else if !exists {
-		return fmt.Errorf("cannot link non-existent blob")
+		return nil, err
 	}
 
-	// The contents of the "link" file are the exact string contents of the
-	// digest, which is specified in that package.
-	return bs.driver.PutContent(bs.ctx, path, []byte(dgst))
+	return p, err
 }
 
-// linked reads the link at path and returns the content.
-func (bs *blobStore) linked(path string) ([]byte, error) {
-	linked, err := bs.readlink(path)
+func (bs *blobStore) Open(ctx context.Context, desc distribution.Descriptor) (distribution.ReadSeekCloser, error) {
+	path, err := bs.path(desc.Digest)
 	if err != nil {
 		return nil, err
 	}
 
-	return bs.get(linked)
+	return newFileReader(ctx, bs.driver, path, desc.Length)
 }
 
-// readlink returns the linked digest at path.
-func (bs *blobStore) readlink(path string) (digest.Digest, error) {
-	content, err := bs.driver.GetContent(bs.ctx, path)
-	if err != nil {
-		return "", err
-	}
-
-	linked, err := digest.ParseDigest(string(content))
+func (bs *blobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, r *http.Request, desc distribution.Descriptor) error {
+	path, err := bs.path(desc.Digest)
 	if err != nil {
-		return "", err
-	}
-
-	if exists, err := bs.exists(linked); err != nil {
-		return "", err
-	} else if !exists {
-		return "", fmt.Errorf("link %q invalid: blob %s does not exist", path, linked)
+		return err
 	}
 
-	return linked, nil
-}
+	redirectURL, err := bs.driver.URLFor(ctx, path, map[string]interface{}{"method": r.Method})
+
+	switch err {
+	case nil:
+		// Redirect to storage URL.
+		http.Redirect(w, r, redirectURL, http.StatusTemporaryRedirect)
+	case driver.ErrUnsupportedMethod:
+		// Fallback to serving the content directly.
+		br, err := bs.Open(ctx, desc)
+		if err != nil {
+			return err
+		}
+		defer br.Close()
 
-// resolve reads the digest link at path and returns the blob store link.
-func (bs *blobStore) resolve(path string) (string, error) {
-	dgst, err := bs.readlink(path)
-	if err != nil {
-		return "", err
+		w.Header().Set("ETag", desc.Digest.String()) // If-None-Match handled by ServeContent
+		w.Header().Set("Cache-Control", fmt.Sprintf("max-age=%d", 86400))
+		w.Header().Set("Content-Type", desc.MediaType)
+		http.ServeContent(w, r, desc.Digest.String(), time.Time{}, br)
```

Comment:
It seems that ```driver.ErrUnsupportedMethod``` is an expected case, so should we really return this error?
If not, ```return nil``` must be there. Otherwise we return an error, which is expected, and it's not proper according to a comment below (line 123)